core/domain/certificate_assessment_services.py

```
@@ -279,6 +302,391 @@ def _has_valid_distinct_assignment(
+        attempt and the question/version payload for the client.
+
+    Raises:
+        CertificateAssessmentAttemptNotReadyException. If the assessment can no
+            longer be started because the question pool is invalid.
+        utils.ValidationError. If the learner already has an in-progress
+            attempt.
+    """
+
+    def _start_txn() -> Tuple[
+        certificate_assessment_domain.CertificateAssessmentAttempt,
+        List[Dict[str, Union[str, int]]],
```

Ruff (0.16.0)

[warning] Use `X | Y` for type annotations

Convert to `X | Y`

(UP007)